x-pack/libbeat/management/fleet/manager.go

```diff
@@ -122,38 +126,51 @@ func (cm *Manager) CheckRawConfig(cfg *common.Config) error {
 	return nil
 }
 
+// UpdateStatus updates the manager with the current status for the beat.
+func (cm *Manager) UpdateStatus(status management.Status, msg string) {
+	cm.lock.Lock()
+	defer cm.lock.Unlock()
+
+	if cm.status != status || cm.msg != msg {
+		cm.status = status
+		cm.msg = msg
+		cm.client.Status(statusToProtoStatus(status), msg)
+		cm.logger.Infof("Status change to %s: %s", status, msg)
+	}
+}
+
 func (cm *Manager) OnConfig(s string) {
-	cm.client.Status(proto.StateObserved_CONFIGURING, "Updating configuration")
+	cm.UpdateStatus(management.Configuring, "Updating configuration")
 
 	var configMap common.MapStr
 	uconfig, err := common.NewConfigFrom(s)
 	if err != nil {
 		err = errors.Wrap(err, "config blocks unsuccessfully generated")
 		cm.logger.Error(err)
-		cm.client.Status(proto.StateObserved_FAILED, err.Error())
+		cm.UpdateStatus(management.Failed, err.Error())
 		return
 	}
 
 	err = uconfig.Unpack(&configMap)
 	if err != nil {
 		err = errors.Wrap(err, "config blocks unsuccessfully generated")
 		cm.logger.Error(err)
-		cm.client.Status(proto.StateObserved_FAILED, err.Error())
+		cm.UpdateStatus(management.Failed, err.Error())
 		return
 	}
 
 	blocks, err := cm.toConfigBlocks(configMap)
 	if err != nil {
 		err = errors.Wrap(err, "could not apply the configuration")
 		cm.logger.Error(err)
-		cm.client.Status(proto.StateObserved_FAILED, err.Error())
+		cm.UpdateStatus(management.Failed, err.Error())
```

Comment:
given all these errors I wonder what 'Failed' means. Is the beat really 'failed' or are we (the overall application with agent) running in a degraded mode because not all configurations could have been applied? Even if we fail here the Beat continues to publish events for existing inputs, right?

Reply:
As stated in the list of statuses there is `Failed` and `Degraded`.

```
        // Degraded is status describing application is degraded.
	Degraded
	// Failed is status describing application is failed. This status should
	// only be used in the case the beat should stop running as the failure
	// cannot be recovered.
	Failed
```

In Degarded state beat can keep sending events and agent will not stop the process. `Failed` is failed, I have completely failed, restart me.

Reply:
Yeah, I've seen these status types. Is this method run on startup only, or also if the beat is already active? In the later case the Beat will continue with the old state if there was an error decoding the configuration. Do we want to enforce a 'restart' if the config update operation failed? Will `Failed` trigger a restart by the Agent?

> Are we mixing the status of the update 'operation' with the beat health here? If the operation fails, are we required to restart the Beat?

I'm more or less wondering about the overal semantics of `OnConfig` and the single errors we can get.

Reply:
It will run both on startup and while running.

The beat doesn't need to worry about handling the restart on Failed status, the Agent will do that work for it, restart it and bring it back up.